libnavigation-base/src/main/java/com/mapbox/navigation/base/utils/route/RouteUtils.kt

```diff
@@ -0,0 +1,162 @@
+                if (silentWaypointsInClosureRange.isNotEmpty()) {
+                    silentWaypointsInClosureRange.forEach {
+                        val isSnapAllowed =
+                            snappingResultList.getOrNull(index + 1 + it) ?: false
+                        if (!isSnapAllowed) {
+                            logW(
+                                "Route with id [${navigationRoute.id}] has closure at leg " +
+                                    "index $routeLegIndex, that overlaps silent (via) waypoint",
+                                LOG_CATEGORY
+                            )
+                            return true
+                        }
+                    }
+                }
+                if (closure.geometryIndexStart() != 0 &&
+                    closure.geometryIndexEnd() != legLastGeometryIndex &&
+                    silentWaypointsInClosureRange.isEmpty()
+                ) {
+                    logW(
+                        "Route with id [${navigationRoute.id}] has closure at leg " +
+                            "index $routeLegIndex",
+                    )
+                    return true
+                }
+            }
+            index += silentWaypoints.size
+        }
+
+        return false
+    }
+
+    private fun DirectionsRoute.getSnappingResultList(): List<Boolean> {
+        val snappingIncludeClosuresList = routeOptions()?.snappingIncludeClosuresList()
+        val snappingIncludeStaticClosuresList = routeOptions()?.snappingIncludeStaticClosuresList()
+
+        val snappingResultList = mutableListOf<Boolean>()
+        for (index in 0 until maxOf(
+            snappingIncludeClosuresList?.size ?: 0,
+            snappingIncludeStaticClosuresList?.size ?: 0
+        )
+        ) {
+            snappingResultList.add(
+                snappingIncludeClosuresList?.getOrNull(index) ?: false ||
+                    snappingIncludeStaticClosuresList?.getOrNull(index) ?: false
+            )
+        }
+        return snappingResultList
+    }
+
+    /**
+     * return the list of indexes of List<LegSilentWaypoints> which are at closure area
+     */
+    private fun List<LegSilentWaypoints>.inGeometryRange(
+        geometryIndexStart: Int,
+        geometryIndexEnd: Int
+    ): List<Int> {
+        return mapIndexedNotNull { index, legSilentWaypoints ->
+            if (legSilentWaypoints.geometryIndex in geometryIndexStart..geometryIndexEnd) {
+                index
+            } else {
+                null
+            }
+        }
+    }
+
+    private fun RouteLeg.stepsGeometryToPoints(
+        precision: Int
+    ): List<Point> {
+        val points = mutableSetOf<Point>()
+
+        steps()?.forEach { step ->
+            step.geometry()?.let { geometry ->
+                points.addAll(PolylineUtils.decode(geometry, precision))
+            } ?: run {
+                logE("LegStep doesn't have geometry", LOG_CATEGORY)
+            }
+        }
+        return points.toList()
```

Comment:
This doesn't deduplicate points, so the duplicates at start and ends of steps will inflate the indices. Also, the `DecodeUtils` should be used (on a worker) so that the cache can be reused across the app.